src/Raven.Server/Documents/Handlers/Processors/Streaming/StreamingHandlerProcessorForGetStreamQuery.cs

```diff
@@ -0,0 +1,146 @@
+using Raven.Server.ServerWide;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Handlers.Processors.Streaming
+{
+    internal class StreamingHandlerProcessorForGetStreamQuery : AbstractStreamingHandlerProcessorForGetStreamQuery<DatabaseRequestHandler, DocumentsOperationContext>
+    {
+        protected QueryOperationContext _queryContext;
+
+        public StreamingHandlerProcessorForGetStreamQuery([NotNull] DatabaseRequestHandler requestHandler, HttpMethod method) : base(requestHandler, method)
+        {
+        }
+
+        protected override RequestTimeTracker GetTimeTracker()
+        {
+            return new RequestTimeTracker(HttpContext, Logger, RequestHandler.Database.NotificationCenter, RequestHandler.Database.Configuration, "StreamQuery",
+                doPerformanceHintIfTooLong: false);
+        }
+
+        protected override ValueTask<BlittableJsonReaderObject> GetDocumentData(DocumentsOperationContext context, string fromDocument)
+        {
+            using (context.OpenReadTransaction())
+            {
+                var document = RequestHandler.Database.DocumentsStorage.Get(context, fromDocument);
+                return ValueTask.FromResult(document?.Data.Clone(context));
+            }
+        }
+
+        protected override IDisposable AllocateContext(out DocumentsOperationContext context)
+        {
+            var queryContext = QueryOperationContext.Allocate(RequestHandler.Database);
+            context = queryContext.Documents;
+            _queryContext = queryContext;
+            return queryContext;
+        }
+
+        protected override QueryMetadataCache GetQueryMetadataCache()
+        {
+            return RequestHandler.Database.QueryMetadataCache;
+        }
+
+        protected override async ValueTask ExecuteAndWriteIndexQueryStreamEntriesAsync(DocumentsOperationContext context, IndexQueryServerSide query, string format, string _,
+            string[] propertiesArray, string fileNamePrefix, bool ignoreLimit, bool fromSharded, OperationCancelToken token)
+        {
+            //writes either csv or blittable documents if is shard
+            await using (var writer = GetBlittableQueryResultWriter(format, isDebug: true, context, HttpContext.Response, RequestHandler.ResponseBodyStream(), fromSharded, propertiesArray, fileNamePrefix))
+            {
+                try
+                {
+                    await RequestHandler.Database.QueryRunner.ExecuteStreamIndexEntriesQuery(query, _queryContext, HttpContext.Response, writer, ignoreLimit, token).ConfigureAwait(false);
+                }
+                catch (IndexDoesNotExistException)
+                {
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.NotFound;
+                    await writer.WriteErrorAsync($"Index {query.Metadata.IndexName} does not exist");
+
+                    if (fromSharded)
+                        throw;
```

Comment:
I do want to rethrow in this case because otherwise the error will only be written to stream, and the MergedEnumerator that is reading the streams on the orchestrator can't handle that. That's why I want the orchestrator to get an actual error. But yes I still need to catch this error in the orchestrator and set the same status code there as well. I'm working to have the same error handling for both sharded and non-sharded now.

Reply:
Okay, makes sense. Let's review and verify the error handling in final PR.